src/class.c

```diff
@@ -1205,6 +1205,17 @@ void ClassDeclaration::addLocalClass(ClassDeclarations *aclasses)
     aclasses->push(this);
 }
 
+/***************************************
+ * Return true if class is an InputRange.
+ * This is defined as having the members: empty, front, popFront.
+ */
+bool ClassDeclaration::isInputRange()
+{
+    return (this->search(Loc(), Id::Ffront) &&
+            this->search(Loc(), Id::FpopFront) &&
+            this->search(Loc(), Id::Fempty));
+}
+
```

Comment:
Not sure about foreach, maybe it is intentional? The current behaviour has the effect of detecting ill-defined input ranges by erroring, instead of falling back to (trying) opApply (which will likely not exist in most cases). In any case, there should be only one notion of input range in DMD.

Other range types can be added on demand later, maybe they will be needed for better diagnostics for other range related errors?